File: DynamicProgramming/StringProblems/allConstruct/allConstruct.js
function allConstruct(target, wordBank) {
    if (target === '') {return []}  
    let lst = []
    for (let word of wordBank) {
     if (target.indexOf(word) === 0) {
        let suffix = target.slice(word.length)
          if (allConstruct(suffix, wordBank) === true) {
            lst.push(suffix)
            return lst
          }
     }
    }
    return lst
}

console.log(allConstruct('abcdef', ['ab','abc','cd','def','abcd']))
console.log(allConstruct('skateboard', ['bo','rd','ate','t','ska','sk','boar']))
console.log(allConstruct('', ['cat','dog','snake']))
console.log(allConstruct('fernando', ['F','e','r','n','a','n','d','o']))
console.log(allConstruct('fernando', ['f','e','r','n','a','n','d','o']))

File: DynamicProgramming/allConstruct/allConstruct.py
def allConstruct(target, wordBank):
     
     if target == '':
          return [[]]
     
     result = []

     for word in wordBank:
          if target.startswith(word):
               suffix = target[len(word):]
               suffixWays = allConstruct(suffix, wordBank)
               targetWays = [[word, *way] for way in suffixWays]
               result.extend(targetWays)

     return result

     
print(allConstruct('purple', ['purp', 'p', 'ur', 'le', 'purpl']))
print(allConstruct('abcdef', ['ab', 'abc', 'cd', 'def', 'abcd', 'ef', 'c']))
print(allConstruct('hello', ['cat', 'dog', 'mouse']))
print(allConstruct('', ['cat', 'dog', 'mouse']))
print(allConstruct('aaaaaaaaaaaaaaaaaaaaz', ['a', 'aa', 'aaa', 'aaaa']))

print()

def allConstruct(target, wordBank, memo = {}):
     if target in memo:
          return memo[target]
     
     if target == '':
          return [[]]
     
     result = []

     for word in wordBank:
          if target.startswith(word):
               suffix = target[len(word):]
               suffixWays = allConstruct(suffix, wordBank)
               targetWays = [[word, *way] for way in suffixWays]
               result.extend(targetWays)

     memo[target] = result
     return result

     
print(allConstruct('purple', ['purp', 'p', 'ur', 'le', 'purpl']))
print(allConstruct('abcdef', ['ab', 'abc', 'cd', 'def', 'abcd', 'ef', 'c']))
print(allConstruct('hello', ['cat', 'dog', 'mouse']))
print(allConstruct('', ['cat', 'dog', 'mouse']))
print(allConstruct('aaaaaaaaaaaaaaaaaaaaz', ['a', 'aa', 'aaa', 'aaaa', 'aaaaa']))